chirp/drivers/retevis_ha1g.py

```diff
@@ -0,0 +1,3105 @@
+                return HandShakeStuts.Normal
+            else:
+                return HandShakeStuts.RadioWrong
+        else:
+            return HandShakeStuts.PwdWrong
+    else:
+        return HandShakeStuts.Wrong
+
+
+def handle_connect_ver(
+    current_page_byte: bytes, serial_conn, max_retries=5, chunk_size=1024
+) -> tuple[bool, bytes]:
+    new_bytes = current_page_byte
+    if not new_bytes:
+        return False, new_bytes
+    for _ in range(max_retries):
+        if rdtp_page_data_crc16_ver(new_bytes):
+            return True, new_bytes
+        chunk = serial_conn.read(chunk_size)
+        if not chunk:
+            return False, new_bytes
+        new_bytes += chunk
+    return False, new_bytes
+
+
+def rdtp_page_data_crc16_ver(current_Page_Byte: bytes):
+    byteLen = len(current_Page_Byte)
+    if byteLen <= 13:
+        return False
+
+    pageLen = current_Page_Byte[12] - 6 if current_Page_Byte[12] > 6 else 0
+    pageLen += current_Page_Byte[13]
+
+    if byteLen - 23 < pageLen:
+        return False
+    crcBytes = current_Page_Byte[-3:-1]
+    dataBytes = current_Page_Byte[:-3]
+    crcStr = calculate_crc16(dataBytes, 2)
+    return crcBytes == crcStr
+
+
+def get_handshake_bytes(currentModel):
+    model_str = (
+        currentModel.decode("ascii")
+        if isinstance(currentModel, bytes)
+        else str(currentModel)
+    )
+    data_part = (
+        b"\x00\x00\x00\x00\x01\x00"
+        + f"{model_str}".encode("ascii")
+        + b"\x00" * (41 - len(model_str))
+    )
+
+    # 2. sAssembling handshake data
+    handshake_data = (
+        b"RDTP\x01\x00\x00\x00\x00\x00\x00\x00"
+        + len(data_part).to_bytes(2, "little")
+        + data_part
+    )
+
+    # 3. CRC16
+    crc = calculate_crc16(handshake_data, 2)
+    return handshake_data + crc + b"\xff"
+
+
+def get_read_current_page_bytes(self, item: int, serial, status):
+    page_count = 1
+    page_index = 0
+    item_bytes = b""
+    while page_index < page_count:
+        num = 0
+        success = False
+        while num <= 5:
+            serial.write(Get_ReadItemBytes(item, page_index))
+            data_bytes = serial.read(self.page_len)
+            flag, new_data_bytes = handle_connect_ver(data_bytes, serial)
+            if flag:
+                if new_data_bytes[14] != item:
+                    num += 1
+                    time.sleep(0.05)
+                    continue
+                if page_index == 0:
+                    page_count = new_data_bytes[18] | (new_data_bytes[19] << 8)
+                status.cur += len(new_data_bytes)
+                self.status_fn(status)
+                num = 0
+                page_index += 1
+                item_bytes += new_data_bytes[20:-3]
+                success = True
+                break
+            success = False
+            num += 1
+            time.sleep(0.05)
+        if not success:
+            raise errors.RadioError("down Wrong")
```

Comment:
This is not a proper error message. I assume this is just some sort of "radio reported failure" sort of error condition?